cub/device/dispatch/dispatch_histogram.cuh

```diff
@@ -258,113 +243,116 @@ public:
+      static_assert(cuda::std::is_convertible<CommonT, int>::value,
+                    "The common type of `LevelT` and `SampleT` must be "
+                    "convertible to `int`.");
+      static_assert(cuda::std::is_trivially_copyable<CommonT>::value,
+                    "The common type of `LevelT` and `SampleT` must be "
+                    "trivially copyable.");
+
+      static constexpr bool COMMON_T_IS_BUILTIN_FP =
+        cuda::std::is_floating_point<CommonT>::value;
+
+      // The `is_floating_point` trait is only valid for builtin floats, and
+      // evaluates to false for `__half`.
+      static constexpr bool COMMON_T_IS_HALF_T =
+#if defined(__CUDA_FP16_TYPES_EXIST__)
+        cuda::std::is_same<CommonT, __half>::value;
+#else
+        false;
+#endif
 
-        // Initializer
-        template <typename _LevelT>
-        __host__ __device__ __forceinline__ void Init(
-            int     num_output_levels,  // Number of levels in array
-            _LevelT max_,                // Max sample level (exclusive)
-            _LevelT min_,                // Min sample level (inclusive)
-            _LevelT scale_)              // Bin scaling factor
+      union ScaleT
+      {
+        // Used when CommonT is not floating-point to avoid intermediate
+        // rounding errors (see NVIDIA/cub#489).
+        struct FractionT
+        {
+          CommonT bins;
+          CommonT range;
+        } fraction;
+
+        // Used when CommonT is floating-point as an optimization.
+        CommonT reciprocal;
+      };
+
+      CommonT m_max;   // Max sample level (exclusive)
+      CommonT m_min;   // Min sample level (inclusive)
+      ScaleT  m_scale; // Bin scaling
+
+    public:
+      // Initializer
+      __host__ __device__ __forceinline__ void Init(int    num_levels,
+                                                    LevelT max_,
+                                                    LevelT min_)
+      {
+        m_max = static_cast<CommonT>(max_);
+        m_min = static_cast<CommonT>(min_);
+
+        // Compute scale. Optimize for floating point types by pre-computing
+        // the reciprocal.
+        CUB_IF_CONSTEXPR(COMMON_T_IS_BUILTIN_FP)
         {
-            this->num_bins = num_output_levels - 1;
-            this->max = max_;
-            this->min = min_;
-            this->scale = scale_;
+          m_scale.reciprocal = static_cast<CommonT>(
+            (num_levels - 1) / (m_max - m_min));
         }
-
-        // Initializer (float specialization)
-        __host__ __device__ __forceinline__ void Init(
-            int    num_output_levels,   // Number of levels in array
-            float   max_,                // Max sample level (exclusive)
-            float   min_,                // Min sample level (inclusive)
-            float   scale_)              // Bin scaling factor
+#if defined(__CUDA_FP16_TYPES_EXIST__)
+        CUB_ELSE_IF_CONSTEXPR(COMMON_T_IS_HALF_T)
```

Comment:
I don't think this is equivalent to the previous code. In the case of C++11 you are still going to compile the branch above. This leads to compilation error in the case of `__half`, right?

Reply:
Yep -- switching it back to a static dispatch.